### PR TITLE
Improve dual stack testing

### PR DESCRIPTION
**Pull request by howardjohn**
This PR makes a few changes/additions to dual stack testing.

* Allow the echo client to specify the IP family. This means I can do something like `client http://example.com -6` (just like curl!) to ensure its doing IPv6
* Also report the **source** IP used. This allows us to verify what IP was used for the first hop. Currently, the server reports the client IP, but sincethere are multiple hops, we lose the visibility. See the test case I changed for a very clear example here.
* Changed TestWaypointDNS to do DNS resolution instead of calling the VIP directly (note: this "dns" in the test name is referring to the backend, so the test was valid before.. its just even better now)

A followup PR will be to add a mode into `echo.CallOptions` to automatically call both families, so we don't need to manually do it in tests.

### CODE REVIEW

File: pkg/test/echo/cmd/client/main.go [hunk 1]
```diff
@@ -199,6 +203,13 @@ func getRequest(url string) (*proto.ForwardEchoRequest, error) {
 		NewConnectionPerRequest: newConnectionPerRequest,
 		ForceDNSLookup:          forceDNSLookup,
 	}
+	if v4Only && v6Only {
+		return nil, fmt.Errorf("--v6 and --v6-only are mutually exclusive")
```
Comment (bleggett) [inline pkg/test/echo/cmd/client/main.go:207]:
```suggestion
		return nil, fmt.Errorf("--v4-only and --v6-only are mutually exclusive")
```

File: tests/integration/ambient/waypoint_test.go [hunk 2]
```diff
@@ -383,29 +382,32 @@ func TestWaypointDNS(t *testing.T) {
...
-						Check:  check,
+					t.NewSubTest("v6").Run(func(t framework.TestContext) {
+						src.CallOrFail(t, echo.CallOptions{
+							To:            apps.MockExternal,
+							Address:       apps.MockExternal.Config().DefaultHostHeader,
+							ForceIpFamily: echo.ForceIPFamilyV6,
+							Port:          echo.Port{Name: "http"},
+							Scheme:        scheme.HTTP,
+							Count:         1,
+							// The destination will always get an IPv4 address currently...
```
Comment (bleggett) [inline tests/integration/ambient/waypoint_test.go:406]: Is this a TODO/need an issue ref?

Comment (howardjohn) [reply tests/integration/ambient/waypoint_test.go:406]: Probably so. Ill open up an issue

**Comment by kfaseela** [comment]
cc @leosarra

File: tests/integration/ambient/waypoint_test.go (hunk 2 above)
Comment (howardjohn) [reply tests/integration/ambient/waypoint_test.go:406]: opened up one for each

**Review by jewertow — COMMENTED** [review]
Nice! I will revise the tests that I wrote recently.

**Review by dhawton — APPROVED** [review]

**Pull request merged**